src/test/util/logging.cpp

```diff
@@ -16,16 +16,30 @@ DebugLogHelper::DebugLogHelper(std::string message, MatchFn match)
     m_print_connection = LogInstance().PushBackCallback(
         [this](const std::string& s) {
             if (m_found) return;
-            m_found = s.find(m_message) != std::string::npos && m_match(&s);
+            m_found = s.find(m_message) != std::string::npos && m_match(s);
         });
     noui_test_redirect();
+    m_receiving_log = true;
 }
 
-void DebugLogHelper::check_found()
+DebugLogHelper::~DebugLogHelper()
 {
-    noui_reconnect();
-    LogInstance().DeleteCallback(m_print_connection);
-    if (!m_found && m_match(nullptr)) {
+    StopReceivingLog();
+}
+
+void DebugLogHelper::EndAndThrowIfNotFound()
+{
+    StopReceivingLog();
+    if (!m_found) {
         throw std::runtime_error(strprintf("'%s' not found in debug log\n", m_message));
```

Comment:
I will look into replacing this with `Assert()`.

"exception from destructor in c++" is a topic well covered all over the internet. The compiler produces a warning for that. Here is an example:

```cpp
class A 
{
public:
    ~A() { 
        std::cout << "before throw a\n";
        // warning: '~A' has a non-throwing exception specification but can still throw [-Wexceptions]
        //171 |         throw std::runtime_error("a");
        //    |         ^
        throw std::runtime_error("a");
    } 
};

int main(int, char**)
{
    try { 
        A a;
        std::cout << "before throw b\n";
        throw std::runtime_error("b");
    } catch (const std::exception& e) { 
        std::cout << "catch\n"; // never reached
    } 
    return 0;
}
```

```
$ ./t
before throw b
before throw a
Abort trap (core dumped)
$
```

I will look into improving the commit message as well.

Reply:
Changed to use `std::abort()` in the destructor. Reverted to the original interface:

```cpp
{
    ASSERT_DEBUG_LOG("expected message");
    produce log messages;
}
```

The change was necessary in order to pass the timeout to the "final check or abort" function which was the destructor and there is no way to pass parameters to the destructor. Anyway, now I pass the timeout to the constructor, save it in a member variable and use it in the destructor. With the wait it is now:

```cpp
{
    ASSERT_DEBUG_LOG_WAIT("expected message", 30s);
    produce log messages;
} // the destructor will wait 30s for the message to be logged before aborting
```

Reply:
Marking this as "resolved" because I reverted to the original interface of `ASSERT_DEBUG_LOG()` and changed the exception from the destructor to `std::abort()`.

Let me know if you still have concerns (aka if I should "unresolve" it). Thanks, it looks better now!